File: TeamCode/src/main/cpp/opmodes/auton.cpp
#include "auton.h"

Auton::Auton(
    JNIEnv *p_jni, jobject self,
    PID *pid_x, PID *pid_y, PID *pid_z,
    std::vector<double> path,
    std::vector<AutonAction> actions
) : C_OpMode(p_jni, self) {
    this->pid_x = pid_x;
    this->pid_y = pid_y;
    this->pid_z = pid_z;

    this->path = std::move(path);
    this->actions = std::move(actions);

    this->node_index = 0;
    this->action_index = -1;

    this->robot = nullptr;
}

void Auton::runOpMode() {
    this->robot = new Robot(this);

    this->waitForStart();

    for (auto &action : this->actions) {
        if (action.at_node != -1) continue;
        action.run(action, this);
    }

    auto last = std::chrono::high_resolution_clock::now();

    while (this->opModeIsActive()) {
        auto now = std::chrono::high_resolution_clock::now();
        auto delta = now - last;
        last = now;

        this->robot->odometry->update();
        this->printDebugInfo();

        this->target[0] = this->path[this->node_index * 3];
        this->target[1] = this->path[this->node_index * 3 + 1];
        this->target[2] = this->path[this->node_index * 3 + 2];

        double theta = this->robot->odometry->position[2];

        maths::vec3 errors{
            this->target[0] - (this->robot->odometry->position[0] / ODOMETRY_TICKS_PER_CM),
            this->target[1] - (this->robot->odometry->position[1] / ODOMETRY_TICKS_PER_CM),
            this->target[2] + (theta / M_PI * 180.0)
        };

        this->pid_values[0] = this->pid_x->update(errors[0], delta);
        this->pid_values[1] = this->pid_y->update(errors[1], delta);
        this->pid_values[2] = this->pid_z->update(errors[2], delta);

        maths::vec2 rotated_pid_values{
            std::cos(-theta) * this->pid_values[0] - std::sin(-theta) * this->pid_values[1],
            std::sin(-theta) * this->pid_values[0] + std::cos(-theta) * this->pid_values[1]
        };

        this->pid_values[0] = rotated_pid_values[0];
        this->pid_values[1] = rotated_pid_values[1];

        this->robot->drivetrain->drive(maths::vec3{
            this->pid_values[1], // forward
            this->pid_values[2], // turn
            this->pid_values[0], // strafe
        });

        if (
            std::abs(errors[0]) < AUTON_PID_X_THRESHOLD &&
            std::abs(errors[1]) < AUTON_PID_Y_THRESHOLD &&
            std::abs(errors[2]) < AUTON_PID_TH_THRESHOLD
        ) {
            this->action_index = -1;

            for (int i = 0; i < this->actions.size(); i++) {
                auto &action = this->actions[i];

                if (action.done || action.at_node != this->node_index) {
                    continue;
                }

                action.run(action, this);

                if (!action.done) {
                    this->action_index = i;
                    break;
                }
            }

            if (this->action_index == -1) this->node_index++;

            if (node_index >= this->path.size() / 3) {
                node_index = (int) ((this->path.size() / 3) - 1);
            }
        }

        this->telemetry->update();
    }
}

void Auton::printDebugInfo() {
    this->telemetry->addLine(utils::sprintf(
            "(position) x = %.2f; y = %.2f; θ = %.2f",
            this->robot->odometry->position[0] / ODOMETRY_TICKS_PER_CM,
            this->robot->odometry->position[1] / ODOMETRY_TICKS_PER_CM,
            this->robot->odometry->position[2] / M_PI * 180.0
    ));

    this->telemetry->addLine(utils::sprintf(
            "(target) x = %.2f; y = %.2f; θ = %.2f",
            this->path[this->node_index * 3],
            this->path[this->node_index * 3 + 1],
            this->path[this->node_index * 3 + 2]
    ));

    this->telemetry->addLine();
    this->telemetry->addLine(utils::sprintf("forward_power = %.2f", this->pid_values[1]));
    this->telemetry->addLine(utils::sprintf("strafe_power = %.2f", this->pid_values[0]));
    this->telemetry->addLine(utils::sprintf("turn_power = %.2f", this->pid_values[2]));

#ifndef PRACTICE_BOT
    this->telemetry->addLine();
    this->telemetry->addLine(utils::sprintf("lift_1 power = %.2f", this->robot->lift_1->getPower()));
    this->telemetry->addLine(utils::sprintf("lift_2 power = %.2f", this->robot->lift_2->getPower()));
#endif

    this->telemetry->addLine();
    this->telemetry->addLine(utils::sprintf("node index = %d", this->node_index));

    if (this->action_index != -1) {
        auto name = this->actions[this->action_index].name;
        if (name == nullptr || strlen(name) == 0) name = "< unnamed >";

        this->telemetry->addLine(utils::sprintf("current action = %s", name));
    } else {
        this->telemetry->addLine("current action = < none >");
    }

    this->telemetry->addLine();
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/auton/Auton.java
package org.firstinspires.ftc.teamcode.auton;

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;

import org.firstinspires.ftc.teamcode.Constants;
import org.firstinspires.ftc.teamcode.Robot;
import org.firstinspires.ftc.teamcode.utils.PID;
import org.firstinspires.ftc.teamcode.utils.maths.Axis;
import org.firstinspires.ftc.teamcode.utils.maths.Vec3;

import java.util.Locale;

public class Auton {
    LinearOpMode opMode;

    public Robot robot;
    Vec3 target = new Vec3(0, 0, 0);

    int nodeIndex = 0;
    int actionIndex = -1;
    Vec3 errors = new Vec3(0, 0, 0);
    Vec3 pidValues = new Vec3(0, 0, 0);

    public PID pidX = new PID(1, 0, 0);
    public PID pidY = new PID(1, 0, 0);
    public PID pidZ = new PID(1, 0, 0);

    Vec3[] path;
    AutonAction[] actions;

    public Auton(LinearOpMode opMode, Vec3[] path, AutonAction[] actions) {
        this.opMode = opMode;
        this.path = path;
        this.actions = actions;
    }

    public void runOpMode() {
        this.robot = new Robot(this.opMode);

        this.opMode.waitForStart();

        for (AutonAction action : this.actions) {
            if (action.atNode != -1) continue;
            action.run(this);
        }

        while (this.opMode.opModeIsActive()) {
            this.robot.odometry.update();
            this.printDebugInfo();

            this.target = this.path[this.nodeIndex];
            this.errors.x = this.target.x - this.robot.odometry.position.x / Constants.ODOMETRY_TICKS_PER_CM;
            this.errors.y = this.target.y - this.robot.odometry.position.y / Constants.ODOMETRY_TICKS_PER_CM;
            this.errors.z = this.target.z + this.robot.odometry.position.z / Math.PI * 180.0;

            this.pidValues.x = this.pidX.update(this.errors.x);
            this.pidValues.y = this.pidY.update(this.errors.y);
            this.pidValues.z = this.pidZ.update(this.errors.z);

            this.pidValues.rotate(-this.robot.odometry.position.z, Axis.Z);

            this.robot.drivetrain.drive(
                    this.pidValues.y,
                    this.pidValues.x,
                    this.pidValues.z
            );

            if (
                Math.abs(this.errors.x) < Constants.AUTON_PID_X_THRESHOLD &&
                Math.abs(this.errors.y) < Constants.AUTON_PID_Y_THRESHOLD &&
                Math.abs(this.errors.z) < Constants.AUTON_PID_TH_THRESHOLD
            ) {
                this.actionIndex = -1;

                for (int i = 0; i < this.actions.length; i++) {
                    AutonAction action = this.actions[i];
                    if (action.done || action.atNode != this.nodeIndex) continue;

                    action.run(this);

                    if (!action.done) {
                        this.actionIndex = i;
                        break;
                    }
                }

                if (this.actionIndex == -1) this.nodeIndex++;

                if (this.nodeIndex >= this.path.length) {
                    this.nodeIndex = this.path.length - 1;
                }
            }

            this.opMode.telemetry.update();
        }
    }

    void printDebugInfo() {
        this.opMode.telemetry.addLine(String.format(
                Locale.getDefault(),
                "(position) x = %.2f; y = %.2f; θ = %.2f",
                this.robot.odometry.position.x / Constants.ODOMETRY_TICKS_PER_CM,
                this.robot.odometry.position.y / Constants.ODOMETRY_TICKS_PER_CM,
                this.robot.odometry.position.z / Math.PI * 180.0
        ));

        this.opMode.telemetry.addLine(String.format(
                Locale.getDefault(),
                "(target) x = %.2f; y = %.2f; θ = %.2f",
                this.target.x,
                this.target.y,
                this.target.z
        ));

        this.opMode.telemetry.addLine();

        this.opMode.telemetry.addLine(String.format(Locale.getDefault(), "forward_power = %.2f", this.pidValues.y));
        this.opMode.telemetry.addLine(String.format(Locale.getDefault(), "strafe_power = %.2f", this.pidValues.x));
        this.opMode.telemetry.addLine(String.format(Locale.getDefault(), "turn_power = %.2f", this.pidValues.z));

        this.opMode.telemetry.addLine();
        this.opMode.telemetry.addLine(String.format(Locale.getDefault(), "node_index = %d", this.nodeIndex));

        if (this.actionIndex != -1) {
            String name = this.actions[this.actionIndex].name;
            if (name.isEmpty()) name = "< unnamed >";

            this.opMode.telemetry.addLine(String.format(Locale.getDefault(), "current action = %s", name));
        } else {
            this.opMode.telemetry.addLine("current action = < none >");
        }

        this.opMode.telemetry.addLine();
    }
}
